### PR TITLE
Deprecate AddV

### PR DESCRIPTION
**Pull request by permcody**
If anyone has any comments on the new overload in InputParameters please let me know. I was thinking about adding a new name instead of just overloading the existing routines:
`addCoupledVar()`
`addRequiredCoupledVar()`

This may be a little confusing.
Opinions @tonkmr?

### CODE REVIEW

**Comment by tonkmr** [comment]
I agree that overloading the addCoupledVar and addRequiredCoupledVar might be confusing. One possible option would be 

addCoupledAutoGenVars()
addRequiredCoupledAutoGenVars()

File: framework/include/utils/InputParameters.h [hunk 1]
```diff
@@ -278,19 +278,42 @@ class InputParameters : public Parameters
...
+   *   e.g.
+   *   [./foo]
+   *     ...
+   *     some_base = base_
+   *     some_num  = 5
+   *   [../]
+   *
+   *   # The coupling parameter will be passed this vector: "base_0 base_1 base_2 base_3 base_4"
+   */
+  void addCoupledVar(const std::string &name, const std::string &base_name, const std::string &num_name, const std::string &doc_string);
```
Comment (aeslaughter) [inline framework/include/utils/InputParameters.h]: Do we still want to try to find a new name? If so, how about `addCoupledVarByBase` or `addCoupledVarWithBase` or `addCoupledVarGivenABaseNameAndAnIntergerRepresentingHowManyVariablesToCreate`

**Comment by YaqiWang** [comment]
Not sure if my comments are relevant. If so just ignore me. Possibly this problem can be resolved by parser. If I am understanding correctly, this `addV()` is changing the input parameters and generating few parameters on the fly based on something. It could be done by an application parser, where when the parser noticed something in the input file, it can perform some string substitution and change the content of read input file before input parameter is constructed.

**Comment by permcody** [comment]
That is relevant Yaqi and I haven't forgotten about making the Parser more extensible. This particular change will be handled in MOOSE since it very well may be useful in more than just the phase_field module.

**Comment by friedmud** [comment]
I'd like to rethink this in general.

I think we should promote these "variable arrays" to be first class citizens.  There are enough people that want this functionality that we should stop hacking around at it... and make it right.

How about something like this:

```
[Variables]
  [./grain]
    array = 4
  [../]
[]
```

That would automatically create an ArrayVariable that has 4 entries (i.e. it would create 4 actual variables called grain_0, grain_1, grain_2 and grain_3). (NOTE: I don't want to use the word VECTOR here at all because we might actually support Vector variables (supported by vector shape functions) at some point in the future)

Then the input parameter would still be a normal `addCoupledVar()` thing.  BUT: To get the array of variables you would call:

```c++
const VariableValueArray & coupledValueArray(const std::string & name)
```

The only tricky part is... what do we use for the type of `VariableValueArray`?  Traditionally, we used `std::vector<VariableValue *>` but that requires dereferencing which is ugly... and if we hide the fact that there is a pointer behind a type def... it won't be clear that you NEED to dereference when you're using a `VariableValueArray`.

I propose that we use `std::vector<VariableValueProxy>`... where a `VariableValueProxy` is a `MooseArrayProxy<Real>` and a `MooseArrayProxy` is basically a forwarding object to the "real" `MooseArray`.  Essentially, we hide the dereference behind a class.

In this way we can stop doing the whole "vector coupling" thing all together.

**Comment by permcody** [comment]
This is reasonable and is inline with some of the proposed work in the next LDRD. I think the current patch is probably good enough for now.

**Comment by aeslaughter** [comment]
I also like @friedmud's idea, but I need this sooner rather than later to
update the modules with the new constructors.

**Comment by tonkmr** [comment]
@permcody is there a patch for MARMOT ready to go on this?

**Comment by permcody** [comment]
No - but it doesn't break MARMOT either. This needs to be merged all the
way through, then I'll update Marmot later without needing to do an
integration branch change.

**Comment by tonkmr** [comment]
I'm happy with it, then. If someone wants to merge it. I can do it, but I typically try not to merge changes to framework.

**Pull request merged**